### ISSUE TITLE
GRPO num_iterations=2 never reuses a rollout after an optimizer update with the shipped accumulation setting

### ISSUE THREAD

**Issue by benzaid32** [tenstorrent/tt-blacksmith#731]
## Summary

The shipped Gemma 2 GRPO config sets:

```yaml
num_iterations: 2
gradient_accumulation_steps: 4
```

`train_grpo()` refreshes the old policy and generates a new rollout at the start of every dataloader batch. It then counts each `num_iterations` pass as one gradient-accumulation step and only updates the policy after four such passes.

With the shipped `2/4` combination, the schedule is:

| Rollout batch | Iteration | Accumulation count | Policy updated before loss? |
|---|---:|---:|---|
| 1 | 1 | 1 | No |
| 1 | 2 | 2 | No |
| 2 | 1 | 3 | No |
| 2 | 2 | 4 | No; optimizer runs after this loss |
| 3 | 1 | 1 | No; old policy was just refreshed |

The second pass over a rollout is therefore another backward pass against unchanged policy weights, not a second policy update. No loss for that rollout is evaluated after `optimizer.step()`.

This contradicts the documented behavior that later iterations reuse fixed rollouts after the new policy has moved, making the new/old importance ratio and PPO clipping active.

## Affected revision

```text
c1f9378d5d34feb7835e6717f07a6da1d6fd6b5b
```

## Affected code

The shipped settings are in:

```text
blacksmith/experiments/torch/gemma2/grpo/single_chip/gemma2_gsm8k_grpo.yaml
```

```yaml
num_iterations: 2
...
gradient_accumulation_steps: 4
```

The old policy is refreshed once per dataloader batch, before generation:

```python
for batch in progress_bar:
    ...
    sync_old_policy(policy_model)
    ...
    completion_ids, completion_valid = generate_completions(...)
```

Each GRPO iteration increments the same accumulation counter. The optimizer only runs after that counter reaches `gradient_accumulation_steps`:

```python
for _grpo_iter in range(num_iterations):
    logps = forward_logps(policy_model, seq_ids, seq_attention_mask)
    loss, loss_metrics = compute_grpo_loss(..., old_logps=old_logps)

    (loss / config.gradient_accumulation_steps).backward()
    accumulation_step += 1

    if accumulation_step < config.gradient_accumulation_steps:
        continue

    device_manager.optimizer_step(optimizer)
    optimizer.zero_grad()
    accumulation_step = 0
```

The repository documentation states the intended behavior:

```text
for num_iterations updates, recompute only new-policy log-probs on the same sequences
...
With num_iterations>1, later updates reuse the same rollouts against the frozen old policy,
so the importance ratio and PPO clipping become active as the new policy drifts.
```

Source links:

- [shipped config](https://github.com/tenstorrent/tt-blacksmith/blob/c1f9378d5d34feb7835e6717f07a6da1d6fd6b5b/blacksmith/experiments/torch/gemma2/grpo/single_chip/gemma2_gsm8k_grpo.yaml#L12-L28)
- [old-policy refresh and rollout generation](https://github.com/tenstorrent/tt-blacksmith/blob/c1f9378d5d34feb7835e6717f07a6da1d6fd6b5b/blacksmith/experiments/torch/gemma2/grpo/train.py#L194-L231)
- [iteration and accumulation loop](https://github.com/tenstorrent/tt-blacksmith/blob/c1f9378d5d34feb7835e6717f07a6da1d6fd6b5b/blacksmith/experiments/torch/gemma2/grpo/train.py#L269-L315)
- [documented multi-iteration behavior](https://github.com/tenstorrent/tt-blacksmith/blob/c1f9378d5d34feb7835e6717f07a6da1d6fd6b5b/blacksmith/experiments/torch/gemma2/grpo/README.md#L24-L52)

## Reproduction

The attached CPU proof executes the current repository's real `train_grpo()` control flow, GRPO helpers, `AdamW` optimizer, and loss. It replaces only model downloads, dataset I/O, logging, and checkpoints with a deterministic tiny local policy and data. The proof uses the separate old-policy model branch; both that branch and the shipped LoRA-adapter branch share the same refresh, iteration, accumulation, and optimizer ordering shown above.

Run from the workspace used for this report:

```powershell
.\tools\python312\python.exe .\findings\tenstorrent\tt_blacksmith_grpo_num_iterations_accumulation_poc.py
```

Or point the proof at any clone:

```bash
python tt_blacksmith_grpo_num_iterations_accumulation_poc.py /path/to/tt-blacksmith
```

Observed output:

```text
tt-blacksmith commit: c1f9378d5d34feb7835e6717f07a6da1d6fd6b5b
shipped num_iterations: 2
shipped gradient_accumulation_steps: 4

shipped 2/4 schedule ratios:
  loss 1: [1.0, 1.0]
  loss 2: [1.0, 1.0]
  loss 3: [1.0, 1.0]
  loss 4: [1.0, 1.0]
  loss 5: [1.0, 1.0]
  loss 6: [1.0, 1.0]
  loss 7: [1.0, 1.0]
  loss 8: [1.0, 1.0]
shipped clip fractions: [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
shipped optimizer steps: 2
shipped max |ratio - 1|: 0.000000000

control schedule (accumulation=1) ratios:
  loss 1: [1.0, 1.0]
  loss 2: [0.4542405903339386, 1.2347540855407715]
control clip fractions: [0.0, 1.0]
control optimizer steps: 2
control max |ratio - 1|: 0.545759410

PASS: the shipped schedule performs optimizer steps but never evaluates a reused rollout after one
```

The control uses the same loop, policy, and loss but changes `gradient_accumulation_steps` to `1`; each run is supplied enough batches to reach two optimizer steps. It shows the intended second iteration: the first iteration updates the policy, then the second iteration evaluates the same rollout against the frozen old policy and activates clipping.

## Impact

For the shipped `num_iterations=2, gradient_accumulation_steps=4` workload:

- `num_iterations=2` does not perform two policy updates on each rollout.
- Both passes over a rollout use the same policy parameters.
- In the deterministic case, `exp(logps - old_logps)` remains `1` and PPO clipping is inactive.
- The repeated pre-update backward pass is counted toward gradient accumulation, so each optimizer step uses two pre-update gradient evaluations from each of two rollout batches rather than four distinct rollout batches or two updates over a fixed rollout buffer.
- Training completes without an error, so the configured algorithm and the documented behavior silently diverge.

This report does not claim a measured end-to-end accuracy regression. The proven impact is that the shipped training loop does not execute the configured multi-iteration GRPO/PPO update schedule.

## Suggested fix

Keep `num_iterations` and gradient accumulation as separate loop dimensions:

1. Generate and retain the rollout microbatches required for one effective batch while the old policy is frozen.
2. For each of the `num_iterations` policy updates, accumulate gradients over those retained microbatches.
3. Run one optimizer step after each complete accumulation pass.
4. Refresh the old policy and generate new rollouts only after all policy updates for the retained rollout batch finish.

Add a regression test for `num_iterations=2, gradient_accumulation_steps=4` that records policy/old-policy versions at every loss evaluation and verifies:

- two optimizer steps occur before the old policy is refreshed;
- the second update evaluates the same rollout after policy parameters changed;
- a nonzero first update can make the new/old ratio differ from `1` and activate clipping.

If rollout buffering is not supported yet, reject this configuration or set `num_iterations=1` instead of silently presenting repeated backward passes as multiple policy updates.

## Verification limits

- No production service or external system was tested.
- No full Gemma 2 or Tenstorrent hardware run was claimed.
- The proof executes the production training control flow on CPU with deterministic local boundary replacements so the update ordering can be observed directly.

[tt_blacksmith_grpo_num_iterations_accumulation_poc.zip](https://github.com/user-attachments/files/31557865/tt_blacksmith_grpo_num_iterations_accumulation_poc.zip)
